src/lib/services/client-metrics/metrics-service-v2.ts

```diff
@@ -8,47 +8,64 @@ import {
     IClientMetricsStoreV2,
 } from '../../types/stores/client-metrics-store-v2';
 import { clientMetricsSchema } from './schema';
-import { hoursToMilliseconds, minutesToMilliseconds } from 'date-fns';
+import { hoursToMilliseconds, secondsToMilliseconds } from 'date-fns';
 import { IFeatureToggleStore } from '../../types/stores/feature-toggle-store';
 import EventEmitter from 'events';
 import { CLIENT_METRICS } from '../../types/events';
 import ApiUser from '../../types/api-user';
 import { ALL } from '../../types/models/api-token';
 import User from '../../types/user';
+import { collapseHourlyMetrics } from '../../util/collapseHourlyMetrics';
+import { IExperimentalOptions } from '../../experimental';
 
 export default class ClientMetricsServiceV2 {
-    private timer: NodeJS.Timeout;
+    private timers: NodeJS.Timeout[] = [];
+
+    private unsavedMetrics: IClientMetricsEnv[] = [];
 
     private clientMetricsStoreV2: IClientMetricsStoreV2;
 
     private featureToggleStore: IFeatureToggleStore;
 
+    private experimental: IExperimentalOptions;
+
     private eventBus: EventEmitter;
 
     private logger: Logger;
 
-    private bulkInterval: number;
-
     constructor(
         {
             featureToggleStore,
             clientMetricsStoreV2,
         }: Pick<IUnleashStores, 'featureToggleStore' | 'clientMetricsStoreV2'>,
-        { eventBus, getLogger }: Pick<IUnleashConfig, 'eventBus' | 'getLogger'>,
-        bulkInterval = minutesToMilliseconds(5),
+        {
+            experimental,
+            eventBus,
+            getLogger,
+        }: Pick<IUnleashConfig, 'eventBus' | 'getLogger' | 'experimental'>,
+        bulkInterval = secondsToMilliseconds(5),
     ) {
         this.featureToggleStore = featureToggleStore;
         this.clientMetricsStoreV2 = clientMetricsStoreV2;
+        this.experimental = experimental;
         this.eventBus = eventBus;
         this.logger = getLogger(
             '/services/client-metrics/client-metrics-service-v2.ts',
         );
 
-        this.bulkInterval = bulkInterval;
-        this.timer = setInterval(async () => {
-            await this.clientMetricsStoreV2.clearMetrics(48);
-        }, hoursToMilliseconds(12));
-        this.timer.unref();
+        if (this.experimental.batchMetrics) {
+            this.timers.push(
+                setInterval(() => {
+                    this.bulkAdd().catch(console.error);
+                }, bulkInterval).unref(),
+            );
+        }
+
+        this.timers.push(
+            setInterval(() => {
+                this.clientMetricsStoreV2.clearMetrics(48).catch(console.error);
+            }, hoursToMilliseconds(12)).unref(),
+        );
```

Comment:
Should this be part of the else statement?

Reply:
Nope, this is a separate timer for clearing old metrics.